Sources/SwiftSyntax/Syntax.swift

```diff
@@ -372,56 +327,154 @@ extension Syntax {
   }
 }
 
-/// Temporary non-owning Syntax.
-///
-/// This can be used for handling Syntax node without ARC traffic.
-struct UnownedSyntax {
-  private let raw: RawSyntax
-  private let info: Unmanaged<Syntax.Info>
+typealias SyntaxDataReference = SyntaxArenaAllocatedPointer<SyntaxData>
+typealias SyntaxDataReferenceBuffer = SyntaxArenaAllocatedBufferPointer<SyntaxDataReference?>
 
-  @_transparent
-  init(_ node: __shared Syntax) {
-    self.raw = node.raw
-    self.info = .passUnretained(node.info.unsafelyUnwrapped)
+/// Node data for a `Syntax`, allocated and managed by `SyntaxDataArena`.
+/// NOTE: This type must be trivial as it is allocated by ‘BumpPtrAllocator’.
+struct SyntaxData: Sendable {
+  /// Underlying "green" node of this node
+  let raw: RawSyntax
+  /// Reference to the parent data. Or `nil` if this is the root.
+  let parent: SyntaxDataReference?
+  /// Index and position info in the tree.
+  let absoluteInfo: AbsoluteSyntaxInfo
+  /// Length of the children layout.
+  /// This is a cached information, equals to `raw.layoutView?.children.count ?? 0`.
+  let childCount: UInt32
+
+  // If the childCount > 0, the layout buffer (`SyntaxDataArenaReference? * childCount`) is tail allocated.
+}
+
+/// `SyntaxDataArena` manages the entire data of a "red" tree.
+final class SyntaxDataArena: @unchecked Sendable {
+  /// Mutex for locking the data when populating layout buffers.
+  private let mutex: PlatformMutex
+
+  /// Allocator.
+  private let allocator: BumpPtrAllocator
+
+  /// Retaining reference to the arena of the _root_ ``RawSyntax``
+  private let rawArena: RetainedSyntaxArena
+
+  /// Root node.
+  let root: SyntaxDataReference
+
+  init(raw: RawSyntax, rawNodeArena: RetainedSyntaxArena) {
+    precondition(rawNodeArena == raw.arenaReference)
+    assert(MemoryLayout<SyntaxData>.alignment >= MemoryLayout<UnsafePointer<SyntaxData>?>.alignment)
+
+    self.mutex = PlatformMutex.create()
+    self.allocator = BumpPtrAllocator(initialSlabSize: 4096)
```

Comment:
4096 bytes can hold around 100 nodes. We might want to adjust this depending on `raw.totalNodes`